### ISSUE TITLE
Is `bf16-true` precision in FSDP actually mixed precision?

### ISSUE THREAD

**Issue by konstantinjdobler** [Lightning-AI/lightning#18717]
### Bug description

I expected `bf16-true` precision to mean that all weights, gradients, etc. use `bf16`.
However, it turns out that when using FSDP, `bf16-true` precision actually keeps around a master copy of the weights in `fp32` for the optimizer steps. Doing the forward pass in `bf16` but then doing the optimizer step in `fp32` is what I normally would expect from a mixed precision setting. 

The critical lines are here:
https://github.com/Lightning-AI/lightning/blob/b441b3ea2112ba8def3e7f66fd30fcfbabf92f00/src/lightning/pytorch/plugins/precision/fsdp.py#L102-L116

Excerpt from the FSDP MixedPrecision [docs](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.MixedPrecision): "This specifies the dtype for model parameters, inputs [...], and therefore the dtype for computation. However, outside the forward and backward passes, parameters are in full precision. Model checkpointing always happens in full precision." I would have expected to leave MixedPrecision set to `None` when training with `bf16-true`.

Am I missing something? 

Naturally, the main motivation behind `bf16-true` is to save GPU RAM, which is hampered by an extra master copy in `fp32`. But right now, it seems that `lightning` does not natively support true half precision training with FSDP at all. My Llama-7b checkpoint trained with `fabric` in `bf16-true` is ~27GB, which is 4 bytes (`fp32`) per param, so there is definitely happening a conversion to `fp32` somewhere.

### What version are you seeing the problem on?

master

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

cc @carmocca @justusschock @awaelchli

**Comment by awaelchli** [comment]
Hey @konstantinjdobler 
Thank you for reaching out and testing our FSDP implementation in Fabric 🎉 

I can look into the issue you saw with lit-gpt if you give me the command you ran. However, if I just run a simple FSDP example here and inspect the dtypes, everything looks right. 

```py
import torch
import torch.nn as nn
import torch.nn.functional as F

import lightning as L
from lightning.fabric.strategies import FSDPStrategy
from lightning.pytorch.demos import Transformer, WikiText2

policy = {nn.TransformerEncoderLayer, nn.TransformerDecoderLayer}
strategy = FSDPStrategy(auto_wrap_policy=policy, cpu_offload=False)
fabric = L.Fabric(
    accelerator="cuda", 
    devices=1, 
    strategy=strategy,
    precision="bf16-true",
)
fabric.launch()

fabric.seed_everything(42)

with fabric.rank_zero_first():
    dataset = WikiText2()

# 1B parameters
with fabric.init_module():
    model = Transformer(vocab_size=dataset.vocab_size, nlayers=32, nhid=4096, ninp=1024, nhead=64)
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)

num_params = sum([p.numel() for p in model.parameters()])
print(f"expected bf16 memory usage from params: {num_params * 2 / 1e9:.2f} GB")
model, optimizer = fabric.setup(model, optimizer)
print(f"current memory usage with model on device {torch.cuda.max_memory_allocated() / 1e9:.2f} GB")

for i in range(1):
    input, target = fabric.to_device(dataset[i])
    output = model(input.unsqueeze(0), target.unsqueeze(0))
    loss = F.nll_loss(output, target.view(-1))
    fabric.backward(loss)
    print(next(model.parameters()).dtype)
    optimizer.step()
    optimizer.zero_grad()
    print(next(model.parameters()).dtype)
    print(optimizer.param_groups[0]["params"][0].dtype)
    fabric.print(loss.item())

fabric.print(torch.cuda.memory_summary())

fabric.save("bf16-checkpoint.ckpt", {"model": model})
```
Here I'm loading a 1B model, which means it occupies 4 GB of memory in full precision. Switching to `precision=bf16-true` we would expect it to occupy ~2GB of memory, and that's what we're seeing here (see my print statements). Also, during the training loop below, I'm inspecting the dtype of both the model weights and the optimizer parameters outside forward, and I think everything is looking as expected. Finally, I'm saving the checkpoint to a file that ends up being 2GB, so also that looks good. 

There were a few fixes in the past, maybe it got fixed and you were still using an older version. Please let me know if the above makes sense so far, and I'm happy to take a look at the lit-gpt case if you could share me the precise command you ran and lightning + pytorch versions.

**Comment by konstantinjdobler** [comment]
Thank you for trying this out, from your logs it does seem like everything is in order. I will have to try this out in my script, see where the difference is and report back.

**Comment by konstantinjdobler** [comment]
You are indeed correct, `bf16-true` is working as intended. I was tripped by lightning passing a `MixedPrecision` to torch's FSDP even when using `bf16-true` and the `MixedPrecision` docs say "However, outside the forward and backward passes, parameters are in full precision". But when the main model's weights are already in `bf16` this doesn't happen (I guess "full" == `bf16` in this case). I also checked this by modifying the lightning code to not pass `MixedPrecision` to FSDP and with native FSDP without lightning -- same results. 

When passing `Adam(..., foreach=False)` the peak allocated memory also works out to (almost) exactly 4*2.03GB for the params, grads, and optim states (x2), since activations are negligible due to the tiny input.

Thanks again for having a look at this @awaelchli!

**Comment by awaelchli** [comment]
Thank you for checking this. Glad that it worked out. Yes, I think the name MixedPrecision can in some cases be a bit misleading, but at the end of the day, it is just a configuration dataclass. As long as it works, we can treat it as an implementation detail on PyTorch's side and we should be good 🎉 

cheers!
